src/channel/ably-channel.ts

```diff
@@ -44,37 +44,53 @@ export class AblyChannel extends Channel {
     /**
      * Create a new class instance.
      */
-    constructor(ably: any, name: string, options: any) {
+    constructor(ably: any, name: string, options: any, autoSubscribe = true) {
         super();
 
         this.name = name;
         this.ably = ably;
         this.options = options;
         this.eventFormatter = new EventFormatter(this.options.namespace);
         this.subscribedListeners = [];
+        this.errorListeners = [];
+        this.channel = ably.channels.get(name);
 
-        this.subscribe();
+        if (autoSubscribe) {
+            this.subscribe();
+        }
     }
 
     /**
      * Subscribe to an Ably channel.
      */
     subscribe(): any {
-        this.channel = this.ably.channels.get(this.name);
+        this.channel.on(stateChange => {
+            const { previous, current, reason } = stateChange;
+            if (previous !== 'attached' && current == 'attached') {
+                this.subscribedListeners.forEach(listener => listener());
+            } else if (reason) {
+                this._alertErrorListeners(stateChange);
+            }
+        });
+        this.channel.attach(this._alertErrorListeners);
     }
 
     /**
-     * Unsubscribe from an Ably channel.
+     * Unsubscribe from an Ably channel, unregister all callbacks and finally detach the channel
      */
     unsubscribe(): void {
         this.channel.unsubscribe();
+        this.unregisterError();
+        this.unregisterSubscribed();
+        this.channel.off();
+        this.channel.detach();
     }
 
     /**
      * Listen for an event on the channel instance.
      */
-    listen(event: string, callback: Function): AblyChannel {
-        this.on(this.eventFormatter.format(event), callback);
+    listen(event: string, callback: Function | any): AblyChannel {
```

Comment:
We need to keep it this way because internal ably call requires it to be of a different type. As per standard interface we need `callback` to be of type `Function`. We need to make it `any` to make it work for ->  `this.channel.subscribe(this.eventFormatter.format(event), callback);`